### PR TITLE
# Add Contact Message Functionality to API

### PR DESCRIPTION
**Pull request by aaladaruncc**
This PR introduces the functionality for users to send contact messages through the API. The following changes have been made:

- **New Migration**: Added a migration file `0051_contactmessage.py` to create the `ContactMessage` model in the database.
  - The model includes the following fields:
    - `first_name`: CharField, max length of 200
    - `last_name`: CharField, max length of 200
    - `email`: EmailField, max length of 254
    - `message`: TextField
    - `created_at`: DateTimeField, automatically set on creation
  - The model is mapped to the database table `api_contact_message`.

- **Model Implementation**: Defined the `ContactMessage` model in `api/models.py` for storing contact message data.

- **API View**: Created a new API view `ContactPageView` in `api/views/user_views.py` to handle POST requests for sending contact messages.
  - Validates that required fields are present: `first_name`, `last_name`, `email`, and `message`.
  - Includes email format validation to ensure correct structure using Django’s `validate_email` function.
  - If validation is successful, a new `ContactMessage` instance is created and saved to the database.
  - Responds with a success message upon successful creation or relevant error messages for validation failures.

- **URL Routing**: Added a new route in `hice_backend/urls.py` for the contact page with the path `contact-us/`.

- **Unit Tests**: Implemented tests for the `ContactPageView` in `api/tests/tests.py` to verify functionality:
  - Test for successful message sending when all required fields are provided.
  - Testing scenarios for missing fields (`first_name`, `email`, `message`) which should return 400 Bad Request.
  - Test for empty payloads and invalid email formats to ensure proper validation.

These changes enable users to send their messages through the API effectively and ensure robust validation and error handling.

### CODE REVIEW

File: api/models.py [hunk 1]
```diff
@@ -200,3 +200,14 @@ class QuizSessionQuestion(models.Model):
...
         db_table = "api_quiz_session_question"
         unique_together = ("quiz_session", "question")
+
+
+class ContactMessage(models.Model):
+    first_name = models.CharField(max_length=200)
+    last_name = models.CharField(max_length=200)
+    email = models.EmailField()
+    message = models.TextField()
+    created_at = models.DateTimeField(auto_now_add=True)
```
Comment (amhajja) [inline api/models.py:210]: - Let's add a UUID id
- Let's make last_name optional and default it to ''
- `created_at` is a great addition

File: api/views/user_views.py [hunk 2]
```diff
@@ -529,3 +532,34 @@ def post(self, request):
...
+
+
+class ContactPageView(APIView):
+    permission_classes = [AllowAny]
+
+    def post(self, request):
+        fname = request.data.get("first_name")
+        lname = request.data.get("last_name")
+        email = request.data.get("email")
+        message = request.data.get("message")
```
Comment (amhajja) [inline api/views/user_views.py:544]: Let's add a request serializer here, and the decorator that'll allow us to test the endpoint via Swagger

**Review by amhajja — CHANGES_REQUESTED** [review]
Requested a few changes; I like what you did nonetheless!

File: api/views/user_views.py (hunk 2 above)
Comment (aaladaruncc) [reply api/views/user_views.py]: Done

File: api/models.py (hunk 1 above)
Comment (aaladaruncc) [reply api/models.py]: Done

**Review by amhajja — APPROVED** [review]
Very nice; throttling was a nice addition!

**Pull request merged**